File: backend/src/main/java/com/d4k/ecommerce/modules/promotion/service/CouponService.java
package com.d4k.ecommerce.modules.promotion.service;

import com.d4k.ecommerce.modules.promotion.dto.request.ApplyCouponRequest;
import com.d4k.ecommerce.modules.promotion.dto.request.CouponRequest;
import com.d4k.ecommerce.modules.promotion.dto.response.CouponResponse;
import com.d4k.ecommerce.modules.promotion.dto.response.CouponValidationResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Coupon Service Interface
 * Định nghĩa business logic cho coupons
 */
public interface CouponService {
    
    /**
     * Tạo coupon mới (ADMIN)
     * @param request thông tin coupon
     * @return coupon đã tạo
     */
    CouponResponse createCoupon(CouponRequest request);
    
    /**
     * Cập nhật coupon (ADMIN)
     * @param id coupon ID
     * @param request thông tin mới
     * @return coupon đã cập nhật
     */
    CouponResponse updateCoupon(Long id, CouponRequest request);
    
    /**
     * Xóa coupon (ADMIN)
     * @param id coupon ID
     */
    void deleteCoupon(Long id);
    
    /**
     * Lấy chi tiết coupon
     * @param id coupon ID
     * @return coupon detail
     */
    CouponResponse getCouponById(Long id);
    
    /**
     * Lấy danh sách tất cả coupons (ADMIN)
     * @param pageable thông tin phân trang
     * @return danh sách coupons
     */
    Page<CouponResponse> getAllCoupons(Pageable pageable);
    
    /**
     * Lấy danh sách coupons đang valid (PUBLIC)
     * @param pageable thông tin phân trang
     * @return danh sách valid coupons
     */
    Page<CouponResponse> getValidCoupons(Pageable pageable);
    
    /**
     * Search coupons theo keyword (ADMIN)
     * @param keyword từ khóa tìm kiếm
     * @param pageable thông tin phân trang
     * @return danh sách coupons
     */
    Page<CouponResponse> searchCoupons(String keyword, Pageable pageable);
    
    /**
     * Validate và tính toán discount cho coupon
     * @param request coupon code và order amount
     * @return kết quả validation và discount amount
     */
    CouponValidationResponse applyCoupon(ApplyCouponRequest request);
    
    /**
     * Verify coupon code
     * @param code coupon code
     * @return coupon nếu valid
     */
    CouponResponse verifyCouponCode(String code);
}

File: frontend/src/services/coupon-service.js
import apiClient from './api-client';

/**
 * Coupon Service
 * Các API liên quan đến coupons
 */
const couponService = {
  /**
   * Áp dụng coupon code
   * @param {Object} data - { code, orderAmount }
   * @returns {Promise}
   */
  applyCoupon: (data) => {
    return apiClient.post('/coupons/apply', data);
  },

  /**
   * Lấy danh sách coupons available
   * @returns {Promise}
   */
  getAvailableCoupons: () => {
    return apiClient.get('/coupons');
  },

  /**
   * Validate coupon code
   * @param {String} code - Coupon code
   * @returns {Promise}
   */
  validateCoupon: (code) => {
    return apiClient.get(`/coupons/validate/${code}`);
  },
  // ============ ADMIN METHODS ============

  /**
   * Lấy tất cả coupons (Admin)
   * @param {Object} params - Query parameters
   * @returns {Promise}
   */
  getAllCoupons: (params = {}) => {
    return apiClient.get('/admin/coupons', { params });
  },

  /**
   * Tìm kiếm coupons (Admin)
   * @param {String} keyword - Keyword
   * @param {Object} params - Query parameters
   * @returns {Promise}
   */
  searchCoupons: (keyword, params = {}) => {
    return apiClient.get('/admin/coupons/search', {
      params: { keyword, ...params }
    });
  },

  /**
   * Lấy chi tiết coupon (Admin)
   * @param {Number} id - Coupon ID
   * @returns {Promise}
   */
  getCouponById: (id) => {
    return apiClient.get(`/admin/coupons/${id}`);
  },

  /**
   * Tạo coupon mới (Admin)
   * @param {Object} data - Coupon data
   * @returns {Promise}
   */
  createCoupon: (data) => {
    return apiClient.post('/admin/coupons', data);
  },

  /**
   * Cập nhật coupon (Admin)
   * @param {Number} id - Coupon ID
   * @param {Object} data - Coupon data
   * @returns {Promise}
   */
  updateCoupon: (id, data) => {
    return apiClient.put(`/admin/coupons/${id}`, data);
  },

  /**
   * Xóa coupon (Admin)
   * @param {Number} id - Coupon ID
   * @returns {Promise}
   */
  deleteCoupon: (id) => {
    return apiClient.delete(`/admin/coupons/${id}`);
  },
};

export default couponService;

